state/txindex/indexer_service.go

```diff
@@ -31,35 +31,56 @@ func NewIndexerService(idr TxIndexer, eventBus *types.EventBus) *IndexerService
 // OnStart implements cmn.Service by subscribing for all transactions
 // and indexing them by tags.
 func (is *IndexerService) OnStart() error {
-	blockHeadersCh := make(chan interface{})
-	if err := is.eventBus.Subscribe(context.Background(), subscriber, types.EventQueryNewBlockHeader, blockHeadersCh); err != nil {
+	blockHeadersSub, err := is.eventBus.Subscribe(context.Background(), subscriber, types.EventQueryNewBlockHeader)
+	if err != nil {
 		return err
 	}
 
-	txsCh := make(chan interface{})
-	if err := is.eventBus.Subscribe(context.Background(), subscriber, types.EventQueryTx, txsCh); err != nil {
+	// If there're no other subscribers, goroutine below may not pull messages
+	// fast enough. To solve it, we either need to increase the capacity here or
+	// change pubsub to wait a little bit (say, 10ms) before cancelling the
+	// subscription.
+	txsSub, err := is.eventBus.Subscribe(context.Background(), subscriber, types.EventQueryTx, 100)
```

Comment:
Or maybe we should use SubscribeUnbuffered here? This is an internal subscription and we want to provide some level of reliability to our indexing, even if we don't to the realtime web socket subscriptions.

Alternatively, as discussed, maybe we don't need the indexing to be done through the event system - we can just give it the block and results in a go routine once processing is done.

Reply:
Switched to SubscribeUnbuffered